pkgs/test_core/lib/src/runner/reporter/github.dart

```diff
@@ -0,0 +1,196 @@
+  void _onTestStarted(LiveTest liveTest) {
+    // Convert the future to a stream so that the subscription can be paused or
+    // canceled.
+    _subscriptions.add(
+        liveTest.onComplete.asStream().listen((_) => _onComplete(liveTest)));
+
+    // Collect messages from tests as they are emitted.
+    _subscriptions.add(liveTest.onMessage.listen((message) {
+      if (_completedTests.contains(liveTest)) {
+        // The test has already completed and it's previous messages were
+        // written out; ensure this post-completion output is not lost.
+        _sink.writeln(message.text);
+      } else {
+        _testMessages.putIfAbsent(liveTest, () => []).add(message);
+      }
+    }));
+  }
+
+  /// A callback called when [liveTest] finishes running.
+  void _onComplete(LiveTest test) {
+    final errors = test.errors;
+    final messages = _testMessages[test] ?? [];
+    final skipped = test.state.result == Result.skipped;
+    final failed = errors.isNotEmpty;
+    final loadSuite = test.suite is LoadSuite;
+    final synthetic = loadSuite ||
+        test.individualName == setUpAllName ||
+        test.individualName == tearDownAllName;
+
+    // Mark this test as having completed.
+    _completedTests.add(test);
+
+    // Don't emit any info for loadSuite, setUpAll, or tearDownAll tests
+    // unless they contain errors or other info.
+    if (synthetic && (errors.isEmpty && messages.isEmpty)) {
+      return;
+    }
+
+    var defaultIcon =
+        synthetic ? _GithubHelper.synthetic : _GithubHelper.passed;
+    final prefix = failed
+        ? _GithubHelper.failed
+        : skipped
+            ? _GithubHelper.skipped
+            : defaultIcon;
+    final statusSuffix = failed
+        ? ' (failed)'
+        : skipped
+            ? ' (skipped)'
+            : '';
+
+    var name = test.test.name;
+    if (!loadSuite) {
+      if (_printPath && test.suite.path != null) {
+        name = '${test.suite.path}: $name';
+      }
+    }
+    _sink.writeln(_helper.startGroup('$prefix $name$statusSuffix'));
+    for (var message in messages) {
+      _sink.writeln(message.text);
+    }
+    for (var error in errors) {
+      _sink.writeln('${error.error}');
+      _sink.writeln(error.stackTrace.toString().trimRight());
+    }
+    _sink.writeln(_helper.endGroup);
+  }
+
+  void _onDone(bool? success) {
+    _cancel();
+
+    _sink.writeln();
+
+    final hadFailures = _engine.failed.isNotEmpty;
+    final message = StringBuffer('${_engine.passed.length} '
+        '${pluralize('test', _engine.passed.length)} passed');
+    if (_engine.failed.isNotEmpty) {
+      message.write(', ${_engine.failed.length} failed');
+    }
+    if (_engine.skipped.isNotEmpty) {
+      message.write(', ${_engine.skipped.length} skipped');
+    }
+    message.write('.');
+    _sink.writeln(
+      hadFailures
+          ? _helper.error(message.toString())
+          : '${_GithubHelper.success} $message',
+    );
+  }
+}
+
+class _GithubHelper {
```

Comment:
Why not just make this a `abstract` class w/ everything static?

Planning on having state at some point?

Reply:
Or just make these top level private members

Reply:
I'd like to keep these grouped with the other github markup code (`startGroup()`, `error()`, ...) as they're related. No real options other than that; `static` members works for me.